packages/core/src/model.ts

```diff
@@ -6,24 +6,82 @@ import type {
   ExtractEvent,
   EventObject
 } from './types';
+import { mapValues } from './utils';
 
-export interface ContextModel<TContext, TEvent extends EventObject> {
+export interface ContextModel<
+  TContext,
+  TEvent extends EventObject,
+  TEM extends EventCreatorMap<TEvent> = EventCreatorMap<TEvent>
+> {
   initialContext: TContext;
   assign: <TEventType extends TEvent['type'] = TEvent['type']>(
     assigner:
       | Assigner<TContext, ExtractEvent<TEvent, TEventType>>
       | PropertyAssigner<TContext, ExtractEvent<TEvent, TEventType>>,
     eventType?: TEventType
   ) => AssignAction<TContext, ExtractEvent<TEvent, TEventType>>;
+  events: FullEventCreatorMap<TEM>;
+  types: {
+    context: TContext;
+    events: TEvent;
+  };
   reset: () => AssignAction<TContext, any>;
 }
 
+type EventCreatorMap<TEvent extends EventObject> = {
+  [key in TEvent['type']]: (
+    ...args: any[]
+  ) => Omit<TEvent & { type: key }, 'type'>;
+};
+
+type FullEventCreatorMap<
+  TEM extends EventCreatorMap<any>,
+  TE extends EventObject = GetEvents<TEM>
+> = TEM extends any
+  ? {
+      [K in keyof TEM]: (
+        ...args: Parameters<TEM[K]>
+      ) => TE extends { type: K } ? TE : never;
+    }
+  : never;
+
+type Expand<T> = T extends infer O ? { [K in keyof O]: O[K] } : never;
+type Distribute<U, V extends Record<string, () => object>> = U extends keyof V &
+  string
+  ? Expand<{ type: U } & ReturnType<V[U]>>
+  : never;
+
+type GetEvents<T extends EventCreatorMap<any>> = Distribute<keyof T, T>;
```

Comment:
This is quite convoluted right now 🤔 but I also couldn't make it simpler so far in a way that would satisfy the `EventObject` constraints in places where this type is used. A simpler version (which only works for `GetEvents` used on its own):
```ts
type GetEvents<T extends EventCreatorMap<any>> = {
    [K in keyof T]: Expand<{ type: K } & ReturnType<T[K]>> 
}[keyof T]
```
[TS playground](https://www.typescriptlang.org/play?ssl=15&ssc=11&pln=13&pc=1#code/JYOwLgpgTgZghgYwgAgKIDcLgPICMBWECYyA3gFDJXJgCeADhAFzIDOYUoA5uQL7nk6jNJnABhKBDhgA9lACycegB4AKhiwkIAD0ggAJqxGa8hYgD5kAXjKVkAbQDWEWslDJ1osPYDkQiD4AuoEsABR2VAB00XBQXKwscCC09oF2AJTWltgAtsBgahrgyABkZDQMzMjOrrwANMh+lT7mANx8Av5o2vRJ+mqWNqrIOnqGbiAw0MjYyAD85fYA0hPVLjIwMyEzy4HIvMgsIBCYUO2ClcgA4hBgRWCsaiO6WOP3ElKyCkrKSbTmg1s1AcK3cNQ2Hm2qB6fWUpAqjBYKwOZQASrcAK5QECqSpqXYA5B8JzrTaqNIXYSqCDsazXW73R4UYEwGQyMJwFjsTggLiZKyWeGctgcbgNXAsXBsgA2UhA+zqEWQuFiYQlyplcoaCCOGJyuGg-MFyG0tAAXiwfCqzdw4D59nxzOQgA)

Reply:
This doesn't exactly work unless you add `EventObject & ...`:

```ts
type GetEvents<T extends EventCreatorMap<any>> = EventObject &
  {
    [K in keyof T]: Expand<{ type: K } & ReturnType<T[K]>>;
  }[keyof T];
```

But then it shows up weird:

<img width="337" alt="CleanShot 2021-03-13 at 20 59 40@2x" src="https://user-images.githubusercontent.com/1093738/111054980-0d8bbd80-843f-11eb-9b5f-b281cfe53dc3.png">

We can work around this by adding some `Assert<TAsserted, TExpected>` type so that we have `Assert<GetEvents<...>, EventObject>` to clean this up, but at this point, it's almost as complex as the original types.

Let's have this as a refactor target if you want.